### PR TITLE
Update FAO Publications Translator

### PR DESCRIPTION
**Pull request by swifterslb**
- Added places of publication.
- Added subtitle.
- Modified the conditions to determine the language to conform with the new naming practice.
- Modified the method to extract abstract to accommodate multi-paragraph abstracts.
- Used new testing cases.

### CODE REVIEW

File: FAO Publications.js [hunk 1]
```diff
@@ -281,28 +288,83 @@ function doWeb(doc, url) {
 var testCases = [
 	{
 		"type": "web",
-		"url": "http://www.fao.org/documents/card/en/c/204a6894-b554-472e-9a9b-d84404dfcb9e/",
+		"url": "http://www.fao.org/documents/card/zh/c/d8a82e37-695b-4eca-b6f9-53c12758eb8e/",
```
Comment (zuphilip) [inline FAO Publications.js]: Are the automatic test in Scaffold working for you?

File: FAO Publications.js [hunk 2]
```diff
@@ -65,14 +65,20 @@ function scrape(doc, url) {
...
 		//Some variables always appear and appear at the same location in all document pages.
-		//title
-		newItem.title = ZU.xpathText(doc, '//*[@id="headerN1"]/h1');
+		//title: use en dash to connect main title and subtitle (if subtitle exists)
+		var mainTitle = ZU.xpathText(doc, '//*[@id="headerN1"]/h1');
+		var subTitle = ZU.xpathText(doc, '//h4[@class="csc-firstHeader h1"]');
+		if (subTitle == null) {
+			newItem.title = mainTitle;
+		} else {
+			newItem.title = mainTitle + ' – ' + subTitle;
```
[the remark this answers is not in the document]
Comment (swifterslb) [reply FAO Publications.js]: Changed.

File: FAO Publications.js [hunk 3]
```diff
@@ -82,25 +88,26 @@ function scrape(doc, url) {
...
 		newItem.url = pdfUrl.slice(pdfUrl.indexOf('www'));
-		//language: according to the last letter of PDF file name
-		var lang = pdfUrl.charAt(pdfUrl.indexOf('pdf')-2);
-		switch(lang) {
-			case 'a': 
+		//language: according to the last one (old format) or two (new format) letters of PDF file name
+		var lang_old = pdfUrl.charAt(pdfUrl.indexOf('pdf')-2);
+		var lang_new = pdfUrl.slice(pdfUrl.indexOf('pdf')-3, pdfUrl.indexOf('pdf')-1);
+		switch(true) {
+			case ((lang_old == 'a') || (lang_new == 'AR') || (lang_new == 'ar')):
```
[the remark this answers is not in the document]
Comment (swifterslb) [reply FAO Publications.js]: Changed.

File: FAO Publications.js [hunk 4]
```diff
@@ -65,14 +65,20 @@ function scrape(doc, url) {
...
+		var mainTitle = ZU.xpathText(doc, '//*[@id="headerN1"]/h1');
+		var subTitle = ZU.xpathText(doc, '//h4[@class="csc-firstHeader h1"]');
+		if (subTitle == null) {
+			newItem.title = mainTitle;
+		} else {
+			newItem.title = mainTitle + ' – ' + subTitle;
+		}		
 		//abstract
-		newItem.abstractNote = ZU.xpathText(doc, '//*[@id="mainContentN1"]/p');
+		var abs = ZU.xpath(doc, '//div[@id="mainContentN1"]')[0].innerText;
```
[the remark this answers is not in the document]
Comment (swifterslb) [reply FAO Publications.js]: The reason I changed to ZU.path is that there are multiple-paragraph abstracts (e.g. http://www.fao.org/publications/card/en/c/I8852EN and http://www.fao.org/documents/card/en/c/I9069EN ). ZU.xpathText joins all the text together into one paragraph, which I don't want. Any better suggestion?

File: FAO Publications.js [hunk 5]
```diff
@@ -65,14 +65,20 @@ function scrape(doc, url) {
...
+		newItem.abstractNote = abs.slice(abs.indexOf(':')+3);
```
[the remark this answers is not in the document]
Comment (swifterslb) [reply FAO Publications.js]: Test added.

The innerText structure of that @id is: Abstract:\n\nAbstractText (paragraphs in AbstractText is are also divided using \n\n). So .trim() doesn't help here. I've changed to using the index of first \n\n.

File: FAO Publications.js [hunk 6]
```diff
@@ -205,6 +217,10 @@ function scrape(doc, url) {
 		if (newItem.publisher == null) {
 			newItem.publisher = 'FAO';
 		}
+		//If there's no place, use 'Rome, Italy' as place.
+		if (newItem.place == null) {
+			newItem.place = 'Rome, Italy';
+		}
```
[the remark this answers is not in the document]
Comment (swifterslb) [reply FAO Publications.js]: The "Place of publication" field was only added to the web pages in this year. Places for publications before this year aren't indicated. Rome is the HQ of FAO and most of the publications are published here. I feel it's better to use Rome than leave it empty.

File: FAO Publications.js (hunk 1 above)
Comment (swifterslb) [reply FAO Publications.js]: Not quite... When I select an existing test (e.g. populated from the js file) to run or update, Scaffold reports error: "Translation failed: TypeError: ZU.xpath(...)[0] is undefined". So I need to use "New Web" every time a test is needed.

File: FAO Publications.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (zuphilip) [reply FAO Publications.js]: Okay, as a general rule you should make sure that the object is not `null` before you do further calculations like `[0]` or `.innerText`. I suggest to use:
```js
		var abs = doc.getElementById("mainContentN1");
		if (abs) {
			// Take innerText for preserving newlines
			abs = abs.innerText;
			// Delete "Abstract:\n\n" at the beginning
			newItem.abstractNote = abs.slice(abs.indexOf('\n\n'));
		}
```

[the remark this answers is not in the document]
Comment (swifterslb) [reply FAO Publications.js]: Thanks. I've changed according to your suggestion.

File: FAO Publications.js (hunk 1 above)
Comment (swifterslb) [reply FAO Publications.js]: Sorry. This was an issue when this translator was first created (https://github.com/zotero/translators/pull/1409 ). I've added a comment before the testing cases.

**Comment by zuphilip** [comment]
Thank you @swifterslb!

**Comment by swifterslb** [comment]
Hi @zuphilip , there's an interesting problem. With this translator, Zotero (Standalone 5.0.47) fails to grab the first paragraph of the abstract (both one- and multi-paragraph abstracts) into the "Abstract" field with Chrome (66.0.3359.181), although Scaffold testing results are perfectly fine. However, it works very well with Firefox (Developer Edition 61.0b3). 

Also, Chrome and FF handles `\n\n` differently. In Chrome it's just a line break (like one `\n`), while in FF it's proper, a line break with an empty line below. 

What's the possible cause of this and could you suggest how to fix it? Thanks.

**Comment by zuphilip** [comment]
I will have a look at this...

**Comment by zuphilip** [comment]
Continued in #1662.

**Pull request merged**